### ISSUE TITLE
How can a trader apply new hyperopt result file (XXX.json)  to a fixed time interval ?

### ISSUE THREAD

**Issue by JiSeopHyun** [freqtrade/freqtrade#7133]
Yes I understand

## Describe your environment

  * Operating system: Docker in Synology ( Docker - 83cac7be )
  * Python Version: In docker image Ver
  * CCXT version: In docker image Ver
  * Freqtrade Version:  In docker image Ver
  
## Your question

*Ask the question you have not been able to find an answer in the [Documentation](https://www.freqtrade.io/en/latest/)*

https://www.freqtrade.io/en/stable/advanced-setup/#multiple-instances-using-docker

I got the idea from the above document. With 4 containers, 
    #1 is configured for live trading, 
    #2 for dry trading, 
    #3 for download, and 
    #4 for hyperopt execution. 
It can be run manually. However, I want to run this every 4 hours. 
I don't have enough knowledge yet, so I ask for help.

####################################
             docker-compose.yml
####################################

```
version: '3'
services:
  freqtrade1:
    image: freqtradeorg/freqtrade:latest
    # image: freqtradeorg/freqtrade:develop
    # Use plotting image
    # image: freqtradeorg/freqtrade:develop_plot
    # Build step - only needed when additional dependencies are needed
    # build:
    #   context: .
    #   dockerfile: "./docker/Dockerfile.custom"
    restart: always
    container_name: freqtrade1
    volumes:
      - "./user_data:/freqtrade/user_data"
    # Expose api on port 8080 (localhost only)
    # Please read the https://www.freqtrade.io/en/stable/rest-api/ documentation
    # before enabling this.
    ports:
      - 
    # Default command used when running `docker compose up`
    command: >
      trade
      --logfile /freqtrade/user_data/logs/freqtrade1.log
      --db-url sqlite:////freqtrade/user_data/tradesv3_freqtrade1.sqlite
      --config /freqtrade/user_data/config.json
      --config /freqtrade/user_data/config_1.json
      --strategy FSupertrendStrategy
    environment: 
     - TZ=

  freqtrade2:
    image: freqtradeorg/freqtrade:latest
    # image: freqtradeorg/freqtrade:develop
    # Use plotting image
    # image: freqtradeorg/freqtrade:develop_plot
    # Build step - only needed when additional dependencies are needed
    # build:
    #   context: .
    #   dockerfile: "./docker/Dockerfile.custom"
    restart: always
    container_name: freqtrade_dry
    volumes:
      - "./user_data:/freqtrade/user_data"
    # Expose api on port 8080 (localhost only)
    # Please read the https://www.freqtrade.io/en/stable/rest-api/ documentation
    # before enabling this.
    ports:
      - ""
    # Default command used when running `docker compose up`
    command: >
      trade
      --logfile /freqtrade/user_data/logs/freqtrade_dry.log
      --db-url sqlite:////freqtrade/user_data/tradesv3_dry.sqlite
      --config /freqtrade/user_data/config.json
      --config /freqtrade/user_data/config_dry.json
      --strategy SwingHighToSky
    environment: 
     - TZ=

  freqtrade3:
    image: freqtradeorg/freqtrade:latest
    # image: freqtradeorg/freqtrade:develop
    # Use plotting image
    # image: freqtradeorg/freqtrade:develop_plot
    # Build step - only needed when additional dependencies are needed
    # build:
    #   context: .
    #   dockerfile: "./docker/Dockerfile.custom"
    restart: unless-stopped
    container_name: freqtrade_t1
    volumes:
      - "./user_data:/freqtrade/user_data"
    # Expose api on port 8080 (localhost only)
    # Please read the https://www.freqtrade.io/en/stable/rest-api/ documentation
    # before enabling this.
    ports:
      - ""
    # Default command used when running `docker compose up`
    command: >
      download-data
      --config user_data/config.json
      --days 03 
      --timeframe 1h
    environment: 
     - TZ=

  freqtrade4:
    image: freqtradeorg/freqtrade:latest
    # image: freqtradeorg/freqtrade:develop
    # Use plotting image
    # image: freqtradeorg/freqtrade:develop_plot
    # Build step - only needed when additional dependencies are needed
    # build:
    #   context: .
    #   dockerfile: "./docker/Dockerfile.custom"
    restart: unless-stopped
    container_name: freqtrade_t2
    volumes:
      - "./user_data:/freqtrade/user_data"
    # Expose api on port 8080 (localhost only)
    # Please read the https://www.freqtrade.io/en/stable/rest-api/ documentation
    # before enabling this.
    ports:
      - ""
    # Default command used when running `docker compose up`
    command: >
      hyperopt 
      --hyperopt-loss SharpeHyperOptLossDaily 
      --spaces buy sell roi stoploss trailing  
      --strategy FSupertrendStrategy 
      --config user_data/config.json -e 150 
      --timeframe 1h 
######################
      --timerange 20220722-                               I Can not use --days option. //  Is there another way?
######################                           Because this has to change over time.
    environment: 
     - TZ=                 
```

I want to run container #3 every 4 hours, container #4 after container #3 runs 10 minutes later, 
and /reload_config on container #1 after 3 hours.
Is it possible? Or do you have a better idea?

**Comment by xmatthias** [comment]
I don't think this will be possible - at least not in this way.

Docker-compose services are made to be always running - so running `docker-compose up` will start all 4 containers.
The first 2 services (live / dry) will work this way - however downloading will not work (or rather, it'll run continuously - restarting whenever the prior download finished).
The same will happen with the last container (neither download-data nor hyperopt are made to be up constantly - and will stop once finished, causing a container restart).

I guess it might be possible with some docker-compose hackery - but that clearly is beyond any type of support we can provide here. 
Out of the box, docker-compose does not provide any scheduling possibilities - at least not last time i checked.

**Comment by JiSeopHyun** [comment]
I appreciate it. I understand. I'll find another good way. 

More study is required. ㅠㅠ

**Comment by JiSeopHyun** [comment]
Thank you very much and sorry.  Think of it as my passion.  Let's take a
different approach.  No. 3 was automatically terminated due to a lot of
network traffic, and in case 4, it was shutting down directly by looking at
the creation time of the result file.

Thank you~

The question will be deleted.

2022년 7월 26일 (화) 오후 6:23, Matthias ***@***.***>님이 작성:

**Comment by xmatthias** [comment]
Most likely, the best approach is to use something external to schedule this (cron ?) - which will simply start the containers in the sequence you require.

you'll however need to be a bit careful, running 2 bots - as well as downloading data in parallel will most likely get your IP banned pretty quickly by the exchange - so i'm not entirely sure if this is a very good approach.

**Comment by JiSeopHyun** [comment]
Thank you.

I want to run hyperopt periodically.  Any other way?  If the bot is issuing
a certain loss, is it possible to run a high-option by activating the
freqtrade processor to Hander?  No good ideas come to mind.

2022년 7월 26일 (화) 오후 7:24, Matthias ***@***.***>님이 작성:

**Comment by xmatthias** [comment]
in general, I advise against automatic hyperopt reruns - you'll not really know what the end-result will be - and it'll automatically pick the "best" result - but "best" is often subjective, and has (in my experience) often drawbacks, where you look at the result, and would prefer the 2nd/3rd best result.

A manual check of the result before "going live" with it is usually  the safer option.

**Comment by JiSeopHyun** [comment]
Choosing the 2nd or 3rd result means executing the hyperopt 3 times?  I
know that once executed, the results of buy, sell, stop loss, etc. are
saved in strategy file name.json, which is loaded and applied early with
the reload_config bot API command.  The example is as follows.
{
  "strategy_name": "FSupertrendStrategy",
  "params": {
    "buy": {
      "buy_m1": 6,
      "buy_m2": 5,
      "buy_m3": 1,
      "buy_p1": 21,
      "buy_p2": 19,
      "buy_p3": 9
    },
    "sell": {
      "sell_m1": 2,
      "sell_m2": 3,
      "sell_m3": 7,
      "sell_p1": 7,
      "sell_p2": 18,
      "sell_p3": 8
    },
    "protection": {},
    "roi": {
      "0": 0.36,
      "398": 0.21400000000000002,
      "661": 0.074,
      "1795": 0
    },
    "stoploss": {
      "stoploss": -0.337
    },
    "trailing": {
      "trailing_stop": true,
      "trailing_stop_positive": 0.35,
      "trailing_stop_positive_offset": 0.385,
      "trailing_only_offset_is_reached": true
    }
  },
  "ft_stratparam_v": 1,
  "export_time": "2022-07-26 09:27:02.964330+00:00"
}

Otherwise, does hyperopting mean that the above result contains multiple?
If so, where is the content included in the results?

And I don't think that's a question I'm going to ask here, but before I
open up the strategic needs issue.  Is a strategy possible for predicting
candle learning?  The module is TensorFlow.

2022년 7월 26일 (화) 오후 8:16, Matthias ***@***.***>님이 작성:

**Comment by xmatthias** [comment]
Well hyperopt  runs for X epochs - so for each hyperopt run, you'll in reality get X results.
the "best" based on the loss function is written to the file - but if that's the best depends on your expectations, on the loss function, ...

using `hyperopt-list` and `hyperopt-show`, other results can be shown - and will be written to the file by `hyperopt-show` unless specified otherwise.

You can do whatever you like in the strategy - as long as it's somewhat valid python code. 
you might also wanna check out the #6832 work - which will enable ML automatically (including automatic retraining).

**Comment by JiSeopHyun** [comment]
The hyperopt option does not include a -show or -list option.
_________________________________________________________
$docker-compose run --rm  freqtrade hyperopt -help
_________________________________________________________

usage: freqtrade hyperopt [-h] [-v] [--logfile FILE] [-V] [-c PATH] [-d PATH] [--userdir PATH] [-s NAME] [--strategy-path PATH] [--recursive-strategy-search]
                          [-i TIMEFRAME] [--timerange TIMERANGE] [--data-format-ohlcv {json,jsongz,hdf5}] [--max-open-trades INT] [--stake-amount STAKE_AMOUNT]
                          [--fee FLOAT] [-p PAIRS [PAIRS ...]] [--hyperopt-path PATH] [--eps] [--dmmp] [--enable-protections] [--dry-run-wallet DRY_RUN_WALLET]
                          [--timeframe-detail TIMEFRAME_DETAIL] [-e INT]
                          [--spaces {all,buy,sell,roi,stoploss,trailing,protection,default} [{all,buy,sell,roi,stoploss,trailing,protection,default} ...]] [--print-all]
                          [--no-color] [--print-json] [-j JOBS] [--random-state INT] [--min-trades INT] [--hyperopt-loss NAME] [--disable-param-export]
                          [--ignore-missing-spaces]

options:
  -h, --help            show this help message and exit
  -i TIMEFRAME, --timeframe TIMEFRAME
                        Specify timeframe (`1m`, `5m`, `30m`, `1h`, `1d`).
  --timerange TIMERANGE
                        Specify what timerange of data to use.
  --data-format-ohlcv {json,jsongz,hdf5}
                        Storage format for downloaded candle (OHLCV) data. (default: `json`).
  --max-open-trades INT
                        Override the value of the `max_open_trades` configuration setting.
  --stake-amount STAKE_AMOUNT
                        Override the value of the `stake_amount` configuration setting.
  --fee FLOAT           Specify fee ratio. Will be applied twice (on trade entry and exit).
  -p PAIRS [PAIRS ...], --pairs PAIRS [PAIRS ...]
                        Limit command to these pairs. Pairs are space-separated.
  --hyperopt-path PATH  Specify additional lookup path for Hyperopt Loss functions.
  --eps, --enable-position-stacking
                        Allow buying the same pair multiple times (position stacking).
  --dmmp, --disable-max-market-positions
                        Disable applying `max_open_trades` during backtest (same as setting `max_open_trades` to a very high number).
  --enable-protections, --enableprotections
                        Enable protections for backtesting.Will slow backtesting down by a considerable amount, but will include configured protections
  --dry-run-wallet DRY_RUN_WALLET, --starting-balance DRY_RUN_WALLET
                        Starting balance, used for backtesting / hyperopt and dry-runs.
  --timeframe-detail TIMEFRAME_DETAIL
                        Specify detail timeframe for backtesting (`1m`, `5m`, `30m`, `1h`, `1d`).
  -e INT, --epochs INT  Specify number of epochs (default: 100).
  --spaces {all,buy,sell,roi,stoploss,trailing,protection,default} [{all,buy,sell,roi,stoploss,trailing,protection,default} ...]
                        Specify which parameters to hyperopt. Space-separated list.
  --print-all           Print all results, not only the best ones.
  --no-color            Disable colorization of hyperopt results. May be useful if you are redirecting output to a file.
  --print-json          Print output in JSON format.
  -j JOBS, --job-workers JOBS
                        The number of concurrently running jobs for hyperoptimization (hyperopt worker processes). If -1 (default), all CPUs are used, for -2, all CPUs
                        but one are used, etc. If 1 is given, no parallel computing code is used at all.
  --random-state INT    Set random state to some positive integer for reproducible hyperopt results.
  --min-trades INT      Set minimal desired number of trades for evaluations in the hyperopt optimization path (default: 1).
  --hyperopt-loss NAME, --hyperoptloss NAME
                        Specify the class name of the hyperopt loss function class (IHyperOptLoss). Different functions can generate completely different results, since
                        the target for optimization is different. Built-in Hyperopt-loss-functions are: ShortTradeDurHyperOptLoss, OnlyProfitHyperOptLoss,
                        SharpeHyperOptLoss, SharpeHyperOptLossDaily, SortinoHyperOptLoss, SortinoHyperOptLossDaily, CalmarHyperOptLoss, MaxDrawDownHyperOptLoss,
                        MaxDrawDownRelativeHyperOptLoss, ProfitDrawDownHyperOptLoss
  --disable-param-export
                        Disable automatic hyperopt parameter export.
  --ignore-missing-spaces, --ignore-unparameterized-spaces
                        Suppress errors for any requested Hyperopt spaces that do not contain any parameters.

Common arguments:
  -v, --verbose         Verbose mode (-vv for more, -vvv to get all messages).
  --logfile FILE        Log to the file specified. Special values are: 'syslog', 'journald'. See the documentation for more details.
  -V, --version         show program's version number and exit
  -c PATH, --config PATH
                        Specify configuration file (default: `userdir/config.json` or `config.json` whichever exists). Multiple --config options may be used. Can be set
                        to `-` to read config from stdin.
  -d PATH, --datadir PATH
                        Path to directory with historical backtesting data.
  --userdir PATH, --user-data-dir PATH
                        Path to userdata directory.

Strategy arguments:
  -s NAME, --strategy NAME
                        Specify strategy class name which will be used by the bot.
  --strategy-path PATH  Specify additional strategy lookup path.
  --recursive-strategy-search
                        Recursively search for a strategy in the strategies folder.

**Comment by stash86** [comment]
Sigh.... Docs people... docs....

https://www.freqtrade.io/en/latest/hyperopt/#show-details-of-hyperopt-results

**Comment by JiSeopHyun** [comment]
ㅠㅠ Thank you! 

1) docker-compose run --rm  freqtrade hyperopt-list
2) Save the file after selecting the best result
    ->  docker-compose run --rm  freqtrade hyperopt-show -n {x}

**Comment by xmatthias** [comment]
Exactly, that's correct.

**Comment by JiSeopHyun** [comment]
Thank you. After updating the docker image, 3 containers are running without any errors. Just send the reload_config command to telegram once in container 1 (live). The hiperopt option uses "ShortTradeDurHyperOptLoss" . It's been about 14 days since I saw your public program. The strategy changed "SwingHighToSky" to be applicable to futures and added a few indicators. I have applied Live since 3 days after I encountered freqtradf, and it is a small investment, but now it has generated more than 20% return. We are also considering purchasing a membership. take care. JiSeop in South Korea
